app/services/kong/base_service.rb

```
@@ -0,0 +1,131 @@
+      uri = URI.parse("#{@kong_elb}/consumers/#{consumer}/key-auth")
+      req = Net::HTTP::Post.new(uri)
+      request(req, uri)
+    end
+
+    def consumer_signup(user, key_auth)
+      raise 'Missing key auth APIs' if key_auth.empty?
+
+      kong_id, kong_consumer_name, kong_api_key = handle_key_auth_flow(user.consumer.organization,
+                                                                       user.last_name,
+                                                                       key_auth)
+
+      { kong_id: kong_id, kongUsername: kong_consumer_name, token: kong_api_key }
+    end
+
+    private
+
+    def generate_consumer_name(organization, last_name)
+      # Prepend LPB- during test phase to define consumers from LPB
+      "LPB-#{"#{organization}#{last_name}".gsub(/\W/, '')}"
+    end
+
+    def get_or_create_consumer(consumer_name)
+      begin
+        return get_consumer(consumer_name)['id']
+      rescue RuntimeError
+        # continue if consumer not found
+      end
+      create_consumer(consumer_name)['id']
+    end
+
+    def handle_key_auth_flow(org, last_name, key_auth)
+      kong_consumer_name = generate_consumer_name(org, last_name)
+      kong_id = get_or_create_consumer(kong_consumer_name)
+      current_acls = get_acls(kong_id)['data'].collect { |acl| acl['group'] }
+      key_auth.map do |acl|
+        add_acl(kong_id, acl) if current_acls.exclude? acl
+      end
+      kong_api_key = create_key(kong_id)['key']
+      [kong_id, kong_consumer_name, kong_api_key]
+    end
+
+    def request(req, uri)
+      req.basic_auth 'kong_admin', set_kong_password
+      response = @client.start(uri.host, uri.port) do |http|
+        http.request(req)
+      end
+
+      response.tap { |res| res['ok'] = res.is_a? Net::HTTPSuccess }
+      raise 'Failed to retrieve kong consumers list' unless response['ok']
+
+      JSON.parse(response.body) unless response.body.nil?
+    end
+
+    protected
+
+    def set_kong_elb
+      nil
```

makes more sense to me to raise an exception here instead of returning nil.
elb and password being nil will probably make consuming Kong::BaseService not function, but better if it blew up in someone's face that this is an abstract class, not to be used directly, only extended

---

Could even take it a step further and use a gem like https://github.com/openSUSE/abstract_method
Full disclosure, I've never used that gem, just always wanted to check it out

---

[b21e4e3](https://github.com/department-of-veterans-affairs/lighthouse-platform-backend/pull/78/commits/b21e4e373f71711b30f1870f91bdd16c7950cd5f)